File: Golang/TMNG.go
package engine

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"image/png"
	"strings"
	"time"

	"github.com/hajimehoshi/ebiten/v2"
)

var tMNGSignature = [...]byte{0x8a, 0x4d, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a}
var tPNGSignature = [...]byte{0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a}

type TMNGFrame struct {
	PNG   []byte
	Delay time.Duration
}

type TMNGAnimation struct {
	Width             uint32
	Height            uint32
	TicksPerSecond    uint32
	NominalLayerCount uint32
	NominalFrameCount uint32
	NominalPlayTime   uint32
	SimplicityProfile uint32
	Frames            []TMNGFrame
}

type tAdventureMNGAnimation struct {
	animation *TMNGAnimation
	images    []*ebiten.Image
}

func IsMNG(data []byte) bool {
	return len(data) >= len(tMNGSignature) && bytes.Equal(data[:len(tMNGSignature)], tMNGSignature[:])
}

func tMNGFrameDelay(data []byte) (uint32, bool) {
	if len(data) == 0 {
		return 0, false
	}
	offset := 1
	for offset < len(data) && data[offset] != 0 {
		offset++
	}
	offset++
	if offset >= len(data) {
		return 0, false
	}
	hasDelay := data[offset] != 0
	offset += 4
	if !hasDelay || offset+4 > len(data) {
		return 0, false
	}
	return binary.BigEndian.Uint32(data[offset : offset+4]), true
}

func ParseMNG(data []byte) (*TMNGAnimation, error) {
	if !IsMNG(data) {
		return nil, fmt.Errorf("not an MNG image")
	}
	animation := &TMNGAnimation{TicksPerSecond: 100}
	var frame []byte
	var pendingDelay uint32
	hasPendingDelay := false
	inPNG := false
	for offset := len(tMNGSignature); offset < len(data); {
		if len(data)-offset < 12 {
			return nil, fmt.Errorf("truncated MNG chunk")
		}
		dataLength := int(binary.BigEndian.Uint32(data[offset : offset+4]))
		chunkLength := dataLength + 12
		if chunkLength < 12 || chunkLength > len(data)-offset {
			return nil, fmt.Errorf("invalid MNG chunk length")
		}
		chunkType := string(data[offset+4 : offset+8])
		chunkData := data[offset+8 : offset+8+dataLength]
		switch chunkType {
		case "MHDR":
			if len(chunkData) < 28 {
				return nil, fmt.Errorf("invalid MNG header")
			}
			animation.Width = binary.BigEndian.Uint32(chunkData[0:4])
			animation.Height = binary.BigEndian.Uint32(chunkData[4:8])
			if ticks := binary.BigEndian.Uint32(chunkData[8:12]); ticks > 0 {
				animation.TicksPerSecond = ticks
			}
			animation.NominalLayerCount = binary.BigEndian.Uint32(chunkData[12:16])
			animation.NominalFrameCount = binary.BigEndian.Uint32(chunkData[16:20])
			animation.NominalPlayTime = binary.BigEndian.Uint32(chunkData[20:24])
			animation.SimplicityProfile = binary.BigEndian.Uint32(chunkData[24:28])
		case "FRAM":
			pendingDelay, hasPendingDelay = tMNGFrameDelay(chunkData)
		}
		if !inPNG && chunkType == "IHDR" {
			frame = append(frame[:0], tPNGSignature[:]...)
			inPNG = true
		}
		if inPNG {
			frame = append(frame, data[offset:offset+chunkLength]...)
			if chunkType == "IEND" {
				delay := 100 * time.Millisecond
				if hasPendingDelay {
					milliseconds := (uint64(pendingDelay)*1000 + uint64(animation.TicksPerSecond)/2) / uint64(animation.TicksPerSecond)
					if milliseconds > 0 {
						delay = time.Duration(milliseconds) * time.Millisecond
					}
				}
				animation.Frames = append(animation.Frames, TMNGFrame{PNG: append([]byte(nil), frame...), Delay: delay})
				frame = frame[:0]
				hasPendingDelay = false
				inPNG = false
			}
		}
		offset += chunkLength
	}
	if inPNG || len(animation.Frames) == 0 {
		return nil, fmt.Errorf("MNG contains no complete frames")
	}
	return animation, nil
}

func (animation *TMNGAnimation) FrameIndexAt(elapsed time.Duration) int {
	if animation == nil || len(animation.Frames) == 0 {
		return -1
	}
	if len(animation.Frames) == 1 {
		return 0
	}
	var duration time.Duration
	for _, frame := range animation.Frames {
		duration += frame.Delay
	}
	if duration <= 0 {
		return 0
	}
	elapsed %= duration
	if elapsed < 0 {
		elapsed += duration
	}
	for index, frame := range animation.Frames {
		if elapsed < frame.Delay {
			return index
		}
		elapsed -= frame.Delay
	}
	return len(animation.Frames) - 1
}

func FirstMNGPNGFrame(data []byte) ([]byte, error) {
	animation, err := ParseMNG(data)
	if err != nil {
		return nil, err
	}
	return append([]byte(nil), animation.Frames[0].PNG...), nil
}

func (adventure *TAdventure) loadAdventureMNGFrame(name string, elapsed time.Duration) *ebiten.Image {
	if adventure == nil || adventure.files == nil {
		return nil
	}
	name = strings.TrimSpace(strings.ReplaceAll(name, "\\", "/"))
	key := strings.ToLower(name)
	animation := adventure.mngCache[key]
	if animation == nil && !adventure.mngFailures[key] {
		data, _, err := adventure.files.ReadDataFile(name)
		if err != nil {
			return nil
		}
		parsed, err := ParseMNG(data)
		if err == nil {
			animation = &tAdventureMNGAnimation{animation: parsed, images: make([]*ebiten.Image, len(parsed.Frames))}
			for index, frame := range parsed.Frames {
				image, decodeErr := png.Decode(bytes.NewReader(frame.PNG))
				if decodeErr != nil {
					err = decodeErr
					break
				}
				animation.images[index] = ebiten.NewImageFromImage(image)
			}
		}
		if err != nil {
			adventure.mngFailures[key] = true
			return nil
		}
		adventure.mngCache[key] = animation
	}
	if animation == nil {
		return nil
	}
	index := animation.animation.FrameIndexAt(elapsed)
	if index < 0 || index >= len(animation.images) {
		return nil
	}
	return animation.images[index]
}

File: C++/TMng.cpp
#include "TMng.h"

#include <algorithm>
#include <array>
#include <cmath>
#include <cstring>

namespace {
    constexpr std::array<std::uint8_t, 8> mngSignature{0x8a, 0x4d, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};
    constexpr std::array<std::uint8_t, 8> pngSignature{0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};

    std::uint32_t readBigEndian(const std::uint8_t* data) {
        return static_cast<std::uint32_t>(data[0]) << 24 | static_cast<std::uint32_t>(data[1]) << 16 | static_cast<std::uint32_t>(data[2]) << 8 | data[3];
    }

    std::optional<std::uint32_t> frameDelay(const std::uint8_t* data, std::size_t length) {
        if (length == 0) return std::nullopt;
        std::size_t offset = 1;
        while (offset < length && data[offset] != 0) offset++;
        if (offset == length || ++offset >= length) return std::nullopt;
        const bool hasDelay = data[offset++] != 0;
        if (length - offset < 3) return std::nullopt;
        offset += 3;
        if (!hasDelay || length - offset < 4) return std::nullopt;
        return readBigEndian(data + offset);
    }
}

bool TMng::isMNG(const void* content, std::size_t length) {
    if (content == nullptr || length < mngSignature.size()) return false;
    return std::equal(mngSignature.begin(), mngSignature.end(), static_cast<const std::uint8_t*>(content));
}

std::optional<TMng::Animation> TMng::parse(const void* content, std::size_t length) {
    if (!isMNG(content, length)) return std::nullopt;
    const auto* bytes = static_cast<const std::uint8_t*>(content);
    Animation animation;
    std::vector<std::uint8_t> frame;
    std::optional<std::uint32_t> pendingDelay;
    bool inPng = false;
    std::size_t offset = mngSignature.size();
    while (offset < length) {
        if (length - offset < 12) return std::nullopt;
        const std::uint32_t dataLength = readBigEndian(bytes + offset);
        const std::size_t chunkLength = static_cast<std::size_t>(dataLength) + 12;
        if (chunkLength < 12 || chunkLength > length - offset) return std::nullopt;
        const char* type = reinterpret_cast<const char*>(bytes + offset + 4);
        const std::uint8_t* data = bytes + offset + 8;
        if (std::memcmp(type, "MHDR", 4) == 0) {
            if (dataLength < 28) return std::nullopt;
            animation.width = readBigEndian(data);
            animation.height = readBigEndian(data + 4);
            const std::uint32_t ticks = readBigEndian(data + 8);
            if (ticks > 0) animation.ticksPerSecond = ticks;
            animation.nominalLayerCount = readBigEndian(data + 12);
            animation.nominalFrameCount = readBigEndian(data + 16);
            animation.nominalPlayTime = readBigEndian(data + 20);
            animation.simplicityProfile = readBigEndian(data + 24);
        } else if (std::memcmp(type, "FRAM", 4) == 0) pendingDelay = frameDelay(data, dataLength);
        if (!inPng && std::memcmp(type, "IHDR", 4) == 0) {
            frame.assign(pngSignature.begin(), pngSignature.end());
            inPng = true;
        }
        if (inPng) {
            frame.insert(frame.end(), bytes + offset, bytes + offset + chunkLength);
            if (std::memcmp(type, "IEND", 4) == 0) {
                const double milliseconds = pendingDelay.has_value() ? std::round(static_cast<double>(*pendingDelay) * 1000.0 / animation.ticksPerSecond) : 100.0;
                animation.frames.push_back({std::move(frame), static_cast<std::uint32_t>(std::max(0.0, milliseconds))});
                frame.clear();
                pendingDelay.reset();
                inPng = false;
            }
        }
        offset += chunkLength;
    }
    if (inPng || animation.frames.empty()) return std::nullopt;
    return animation;
}

std::vector<std::uint8_t> TMng::firstPngFrame(const void* content, std::size_t length) {
    std::optional<Animation> animation = parse(content, length);
    if (!animation.has_value()) return {};
    return std::move(animation->frames.front().png);
}
